layer/training/runtime/model_trainer.py

```diff
@@ -38,7 +46,75 @@ class TrainContextDataclassMixin:
     train_index: Optional[str] = None
 
 
+cpu_used_temp = None  # pylint: disable=C0103
+start_time_temp = None  # pylint: disable=C0103
+step_value = 0  # pylint: disable=C0103
+
+
+def _get_gpu_metrics(logger: Logger) -> Dict[str, Dict[str, float]]:
+    metrics = {}
+    gpu_present = nvsmi.is_nvidia_smi_on_path() is not None
+    if gpu_present:
+        try:
+            gpus = nvsmi.get_gpus()
+        except subprocess.CalledProcessError:
+            logger.info(
+                "Nvidia driver not running despite nvidia-smi being on the path. No GPU stats collected."
+            )
+        for gpu in gpus:
+            metrics[gpu.id] = {
+                "utilisation": gpu.gpu_util,
+                "mem_utilisation": round(gpu.mem_util, 2),
+                "mem_used": gpu.mem_used,
+                "mem_total": gpu.mem_total,
+                "id": gpu.id,
+                "name": gpu.name,
+            }
+
+    return metrics
+
+
+def _generate_system_metrics_dict(
+    mem_used: float,
+    mem_allocated: float,
+    mem_used_percent: float,
+    cpus_used: float,
+    cpus_available: float,
+    cpu_utilisation_percent: float,
+    gpu_metrics: Dict[Any, Any],
+) -> Dict[str, float]:
+    metrics = {
+        "Memory Used (MB)": mem_used,
+        "Memory Allocated (MB)": mem_allocated,
+        "Memory Utilisation %": mem_used_percent,
+        "CPUs Used": cpus_used,
+        "CPUs Allocated": cpus_available,
+        "CPU Utilisation %": cpu_utilisation_percent,
+    }
+
+    for gpu in gpu_metrics:
+        metrics.update(
```

Comment:
nit: put `gpu_metrics[gpu]["id"]` and `gpu_metrics[gpu]["name"]` just above here to make the statements below slightly cleaner. Also, could use f-strings instead of `format`.